dirty_cat/_fuzzy_join.py

```diff
@@ -178,105 +323,140 @@ def fuzzy_join(
 
     # Warn if presence of missing values
-    if main_table[main_col].isna().any():
+    if main_table[main_cols].isna().any().any():
         warnings.warn(
-            "You are merging on missing values."
-            " The output correspondence will be random or missing."
-            " To avoid unexpected errors you can drop them.",
+            "You are merging on missing values. "
+            "The output correspondence will be random or missing. "
+            "To avoid unexpected errors you can drop them. ",
             UserWarning,
+            stacklevel=2,
         )
 
-    all_cats = pd.concat([main_col_clean, aux_col_clean], axis=0).unique()
+    main_num_cols = main_table[main_cols].select_dtypes(include="number").columns
+    aux_num_cols = aux_table[aux_cols].select_dtypes(include="number").columns
 
-    if encoder is None:
-        enc = HashingVectorizer(analyzer=analyzer, ngram_range=ngram_range)
-    else:
-        enc = encoder
+    # Check if included columns are numeric:
+    any_numeric = len(main_num_cols) != 0
+    # Check if included columns have mixed types:
+    mixed_types = len(main_cols) > len(main_num_cols)
 
-    enc_cv = enc.fit(all_cats)
-    main_enc = enc_cv.transform(main_col_clean)
-    aux_enc = enc_cv.transform(aux_col_clean)
+    if len(main_cols) == 1 and len(aux_cols) == 1 and any_numeric is False:
+        main_cols = main_cols[0]
+        aux_cols = aux_cols[0]
 
-    all_enc = vstack((main_enc, aux_enc))
-
-    tfidf = TfidfTransformer().fit(all_enc)
-    main_enc = tfidf.transform(main_enc)
-    aux_enc = tfidf.transform(aux_enc)
-
-    # Find nearest neighbor using KNN :
-    neigh = NearestNeighbors(n_neighbors=1)
-
-    neigh.fit(aux_enc)
-    distance, neighbors = neigh.kneighbors(main_enc, return_distance=True)
-    idx_closest = np.ravel(neighbors)
+    if numerical_match in ["number"] and any_numeric and not mixed_types:
+        main_enc, aux_enc = _numeric_encoding(
+            main_table, main_num_cols, aux_table, aux_num_cols
+        )
+        idx_closest, matching_score = _nearest_matches(main_enc, aux_enc)
+    elif numerical_match in ["number"] and any_numeric and mixed_types:
+        main_num_enc, aux_num_enc = _numeric_encoding(
+            main_table, main_num_cols, aux_table, aux_num_cols
+        )
+        main_str_cols = list(set(main_cols) - set(main_num_cols))
+        aux_str_cols = list(set(aux_cols) - set(aux_num_cols))
+        main_str_enc, aux_str_enc = _string_encoding(
+            main_table,
+            main_str_cols,
+            aux_table,
+            aux_str_cols,
+            encoder=encoder,
+            analyzer=analyzer,
+            ngram_range=ngram_range,
+        )
+        main_enc = hstack((main_num_enc, main_str_enc), format="csr")
+        aux_enc = hstack((aux_num_enc, aux_str_enc), format="csr")
+        idx_closest, matching_score = _nearest_matches(main_enc, aux_enc)
+        mm_scaler = MinMaxScaler(feature_range=(0, 1))
```

Comment:
We should move `MinMaxScaler` into `_nearest_matches`, because the other branch conditions would still need to have `matching_score` between 0 and 1.

Reply:
Yes, but I am afraid of doing this will make us have a `matching_score` of 1 for the **best** match, not the **exact** one. This is something that we really don't want with strings, because I find it useful to have the perfect string matches as 1 (and brings us easily back to what pandas.merge does).

Reply:
Hmm, that's tricky because this property is correct for strings but not for numerical values. I don't see an easy solution; should we add a parameter for the exact match?

Reply:
You're right, but this gets tricky with numerical and mixed values. A perfect match with numerical values will have a distance of 0, thus a `matching_score` of 1. However, if the distance is higher than 1, `matching_score` becomes negative, which is weird.

Besides, the current implementation for mixed values has 1 for the *best* match and not the *exact* match since we use `MinMaxScaler`, right? So, the `matching_score` parameter behaves differently for string and mixed values.

I think we can solve that by removing `MinMaxScaler` and dividing the distance by its maximum in the `_nearest_matches` function:
- For perfect matches, the distance is 0 => the similarity is 1
- For best matches, the distance > 0 => the similarity < 1
- For the worse case, the distance is 1 => the similarity is 0

WDYT? :)

Reply:
Great, indeed, this is the best solution I think!